presidio-analyzer/analyzer/pattern_recognizer.py

```diff
@@ -73,6 +81,69 @@ def __black_list_to_regex(black_list):
         regex = r"(?:^|(?<= ))(" + '|'.join(black_list) + r")(?:(?= )|$)"
         return Pattern(name="black_list", regex=regex, score=1.0)
 
+    def __calculate_context_similarity(self,
+                                       context_text,
+                                       context_list,
+                                       simplifier):
+        """Context similarity is 1 if there's exact match between a keyword in
+           context_text and any keyword in context_list
+
+        :param context_text a string of the prefix and suffix of the found
+               match
+        :param context_list a list of words considered as context keywords
+        :param simplifier makes the context easier to manage.
+                          It transforms to singular form, remove punctuation,
+                          etc...
+        """
+
+        if context_list is None:
+            return 0
+
+        context_keywords = self.__context_to_keywords(context_text, simplifier)
+        if context_keywords is None:
+            return 0
+
+        similarity = 0.0
+        for context_keyword in context_keywords:
+            if context_keyword in context_list:
+                self.logger.info("Found context keyword '%s'", context_keyword)
+                similarity = 1
+                break
+
+        return similarity
+
+    @staticmethod
+    def __context_to_keywords(context, simplifier):
+        # if no simplifier is defined, just tokenize the text for best effort
+        if simplifier is None:
+            return context.split()
+
+        # if a simplifier is defined, use it.
+        return simplifier.simplify(context)
+
+    @staticmethod
+    def __extract_context(text, start, end):
+        """Extract context for a specified match
```

Comment:
We're doing this multiple times even though the text doesn't necessarily change between recognizers. I think it should be a post-process but let's discuss this offline